src/components/MonthCalendar.tsx

```diff
@@ -59,21 +64,68 @@ const MonthCalendar = ({ events, onFocusDate, focusDate }: MonthCalendarProps):
             return 0;
         };
 
-        const lastCalendarDay = new Date(new Date(firstCalendarDay).setDate(firstCalendarDay.getDate() + gridItems));
-
-        const calendarEvents = getEventsInRange(firstCalendarDay, lastCalendarDay);
-
-        if (calendarEvents.length === 0) return { height: 0, top: 0 };
-
-        const firstEventDate = new Date(calendarEvents[0].start_time);
-        const lastEventDate = new Date(calendarEvents[calendarEvents.length - 1].end_time);
-
-        const top = getGridNumber(new Date(firstEventDate)) * barUnit;
-        const height = (getGridNumber(new Date(lastEventDate)) * barUnit) - top;
-
+        const campaignEvents = events.filter(e => e.campaign.id === campId)
+            .sort((a, b) => {
+                return new Date(a.start_time).getTime() - new Date(b.start_time).getTime();
+            });
+
+        if (campaignEvents.length === 0) {
+            return { height: 0, top: 0 };
+        }
+
+        const firstEventDate = new Date(campaignEvents[0].start_time);
+        const lastEventDate = new Date(campaignEvents[campaignEvents.length - 1].end_time);
+
+        let bottom, top;
+        if (firstEventDate < firstCalendarDay ) {
+            top = 0;
+        }
+        else if (firstEventDate > lastCalendarDay) {
+            top = 100;
+        }
+        else {
+            top = getGridNumber(new Date(firstEventDate)) * barUnit;
+        }
+
+        if (lastEventDate > lastCalendarDay) {
+            bottom = 100;
+        }
+        else if (lastEventDate < firstCalendarDay) {
+            bottom = 0;
+        }
+        else {
+            bottom = getGridNumber(new Date(lastEventDate)) * barUnit;
+        }
+        if (bottom > 100) {
+            bottom = 100;
+        }
+
+        const height = bottom - top;
         return { height, top };
     };
 
+    const getCampColors = (campId?: number) => {
+        const currentCampaign = campaigns.find(c => c.id === campId);
+        if (!currentCampaign?.color) {
+            return { bg: '#d3d3d3', fg: '#00000' };
+        }
```

Comment:
What's the purpose of this?

Reply:
trying to account for a campaign not having a colour? In the future if the user does not set a colour? Or should I assume it will always have a colour?

Reply:
Ok yeah that's valid! I guess I was just being confused by having it inside this function, but my actual confusion came from not understanding the overall structure of things and not from this particular if statement.

If you use a new `getContrastColor()` function instead as discussed in another thread, you should be able to replace this with just doing `campaign.color || '#d3d3d3'` for background and `getContrastColor(campaign.color || '#d3d3d3')` for text.

The `bg`/`fg` object is becoming less and less valuable when you do this, so in the end I think you should probably ditch that construct and replace this entire `getCampaignColor()` function with simpler code calling out to those specific colors.